dulwich/client.py

```diff
@@ -1102,16 +1107,50 @@ def __init__(self, hostname):
 class SubprocessSSHVendor(SSHVendor):
     """SSH vendor that shells out to the local 'ssh' command."""
 
-    def run_command(self, host, command, username=None, port=None):
-        # FIXME: This has no way to deal with passwords..
-        args = ['ssh', '-x']
-        if port is not None:
-            args.extend(['-p', str(port)])
-        if username is not None:
+    def run_command(self, host, command, username=None, port=None,
+                    password=None, key_filename=None):
+        if password and key_filename:
+            raise NotImplementedError(
+                "You can't set passphrase for ssh key with "
+                "SubprocessSSHVendor, use ParamikoSSHVendor instead"
+            )
+
+        if sys.platform == 'win32':
+            args = ['putty.exe', '-ssh']
+            if password:
+                args.extend(['-pw', str(password)])
+        else:
+            if password:
+                try:
+                    subprocess.call(["sshpass"])
+                except OSError as e:
+                    import warnings
+                    warnings.warn(
+                        "You need sshpass for using password option, "
+                        "you can use ssh-add "
+                        "or private key without password "
+                        "or ParamikoSSHVendor"
+                    )
+                    raise e
+                args = ['sshpass', '-p', str(password)]
+            else:
+                args = ['ssh', '-x']
+
+        if port:
+            if sys.platform == 'win32':
```

Comment:
This is assuming that the only ssh binary on Windows is putty, which isn't necessarily true. Putty is also available on *nix. Would it perhaps be possible to create a PuttySSHVendor class?

Reply:
Problem that there is no good multi platform way for obtain the same with just only subprocess. If I right remember, for sending password directly into stdin of ssh we should use fork or threads or asinc solution and most of all default solutions do it. We can use `pexpect` for it, but again, on windows this solution require a [special way](https://pexpect.readthedocs.io/en/stable/overview.html#windows) and it possibly would have a conflict with current `dulwitch` wrapper

In my opinion `paramiko` variant would be the best option for security and amount of features in compare with subprocess variant. But it you don't want to support this, and user already have `sshpass` (or `putty` in case ow windows) and this user don't want to use the key auth or preconfigured ssh config file instead - security is not a problem of `dulwitch` itself...

Reply:
We can, but then we should add the same check of platform into `get_transport_and_path` for using it by default on windows or leave it for user choice. But this would add more complexity only

Again, in my opinion `paramiko` client would be a better multiplatform solution for all cases...

Reply:
I think this would be simpler; initially, we could just check the value of sys.platform to determine the default value for get_ssh_vendor, but we could move to something more complex in the future.